common/scanner.h

```diff
@@ -1,294 +1,444 @@
+                break;
+            case '\0':
+                if (lexer->eof(lexer)) {
+                    return false;
+                }
+                if (last) {
+                    last = 0;
+                } else {
+                    advance(lexer);
+                }
+                break;
+            default:
+                if (scan_identifier(lexer) || last) {
+                    last = 0;
+                } else {
+                    advance(lexer);
+                }
+        }
+    }
+}
+
+bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
+    if (valid_symbols[LEFT_QUOTED_STRING_DELIM] &&
+        (iswlower(lexer->lookahead) || lexer->lookahead == '_' ||
+         lexer->lookahead == '|')) {
+        lexer->result_symbol = LEFT_QUOTED_STRING_DELIM;
+        return scan_left_quoted_string_delimiter(scanner, lexer);
+    }
+    if (valid_symbols[RIGHT_QUOTED_STRING_DELIM] && (lexer->lookahead == '|')) {
+        advance(lexer);
+        lexer->result_symbol = RIGHT_QUOTED_STRING_DELIM;
+        return scan_right_quoted_string_delimiter(scanner, lexer);
+    }
+    if (scanner->in_string && valid_symbols[STRING_DELIM] &&
+        lexer->lookahead == '"') {
+        advance(lexer);
+        scanner->in_string = false;
+        lexer->result_symbol = STRING_DELIM;
+        return true;
+    }
+
+    while (iswspace(lexer->lookahead)) {
+        skip(lexer);
+    }
+
+    if (!scanner->in_string && lexer->lookahead == '#' &&
+        lexer->get_column(lexer) == 0) {
+        advance(lexer);
+
+        while (iswspace(lexer->lookahead)) {
             advance(lexer);
-            return true;
-          }
-          break;
-        case '\'':
-          if (last) last = 0; else advance(lexer);
-          last = scan_character(lexer);
-          break;
-        case '"':
-          if (last) last = 0; else advance(lexer);
-          scan_string(lexer);
-          break;
-        case '{':
-          if (last) last = 0; else advance(lexer);
-          if (lexer->lookahead == '%') {
+        }
+
+        if (!iswdigit(lexer->lookahead)) {
+            return false;
+        }
+        while (iswdigit(lexer->lookahead)) {
             advance(lexer);
-            if (lexer->lookahead == '%') advance(lexer);
-            if (scan_extattrident(lexer)) {
-              while (iswspace(lexer->lookahead)) advance(lexer);
-            } else {
-              break;
-            }
-          }
-          if (scan_quoted_string(lexer)) advance(lexer);
-          break;
-        case '\0':
-          if (lexer->eof(lexer)) return false;
-          if (last) last = 0; else advance(lexer);
-          break;
-        default:
-          if (scan_identifier(lexer) || last) last = 0; else advance(lexer);
-      }
+        }
+
+        while (iswspace(lexer->lookahead)) {
+            advance(lexer);
+        }
+
+        if (lexer->lookahead != '"') {
+            return false;
+        }
+		advance(lexer);
```

Comment:
It caused an infinite loop bug, with a hanging opening quote followed by an eof (iirc) never being advanced and it looping back to this section indefinitely

I have to ask then, are comments in line number directives meant to be parsed?

Reply:
It looked like a mistake because of the weird indentation. There is indeed a bug here, and adding `advance(lexer)` is probably correct, but it's difficult to review if it's combined with other refactoring.